cql3/functions/aggregate_fcts.cc

```diff
@@ -143,11 +143,26 @@ make_internal_scalar_function(sstring name, null_handler nullhandler, Lambda fun
     return make_internal_scalar_function(std::move(name), nullhandler, +func);
 }
 
+// is_same(t1, t2) compares if two values are the same. This is mostly the
+// same as "==", except for one exception: We consider two floating-point
+// NaNs to be equal, whereas the floating point standard considers NaN to
+// not be equal to anything, not even to itself.
+template <typename T>
+static bool is_same(const T& t1, const T& t2) {
+    return t1 == t2;
+}
+template<> bool is_same(const float& t1, const float& t2) {
+    return std::isnan(t1) ? std::isnan(t2) : t1 == t2;
+}
+template<> bool is_same(const double& t1, const double& t2) {
+    return std::isnan(t1) ? std::isnan(t2) : t1 == t2;
+}
+
 template<typename NarrowT, typename WideT>
 NarrowT
 narrow(WideT acc) {
     NarrowT ret = static_cast<NarrowT>(acc);
-    if (static_cast<WideT>(ret) != acc) {
+    if (!is_same(static_cast<WideT>(ret), acc)) {
```

Comment:
IMO we should just check if NarrowT == WideT. If the two types are different, then t1 == t2 will fail on the least significant digits.

The correct check for float types is not t1 == t2, but t1 != inf && t2 == inf, indicating an overflow. Or maybe we don't need to check anything, since inf is a reasonable result for overflow, better than failing the query.